lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -479,26 +482,27 @@ def do_3d_projection(artist):
                     "%(since)s and will be removed %(removal)s.")
                 return artist.do_3d_projection(renderer)
 
+            collections_and_patches = (
+                artist for artist in self._children
+                if isinstance(artist, (mcoll.Collection, mpatches.Patch)))
             if self.computed_zorder:
                 # Calculate projection of collections and patches and zorder
                 # them. Make sure they are drawn above the grids.
                 zorder_offset = max(axis.get_zorder()
                                     for axis in self._get_axis_list()) + 1
-                for i, col in enumerate(
-                        sorted(self.collections,
-                               key=do_3d_projection,
-                               reverse=True)):
-                    col.zorder = zorder_offset + i
-                for i, patch in enumerate(
-                        sorted(self.patches,
-                               key=do_3d_projection,
-                               reverse=True)):
-                    patch.zorder = zorder_offset + i
+                collection_count = patch_count = 0
+                for artist in sorted(collections_and_patches,
+                                     key=do_3d_projection,
+                                     reverse=True):
+                    if isinstance(artist, mcoll.Collection):
+                        artist.zorder = zorder_offset + collection_count
+                        collection_count += 1
+                    elif isinstance(artist, mpatches.Patch):
+                        artist.zorder = zorder_offset + patch_count
+                        patch_count += 1
```

Comment:
Let's track zorder directly not artist counts

```suggestion
                collection_zorder = patch_zorder = zorder_offset
                for artist in sorted(collections_and_patches,
                                     key=do_3d_projection,
                                     reverse=True):
                    if isinstance(artist, mcoll.Collection):
                        artist.zorder = collection_zorder
                        collection_zorder += 1
                    elif isinstance(artist, mpatches.Patch):
                        artist.zorder = patch_zorder
                        patch_zorder += 1
```